engine/src/main/java/com/google/android/fhir/sync/FhirSyncWorker.kt

```diff
@@ -23,19 +23,26 @@ import com.google.android.fhir.FhirEngine
 import com.google.android.fhir.sync.Result.Success
 
 /** A WorkManager Worker that handles periodic sync. */
-abstract class PeriodicSyncWorker(appContext: Context, workerParams: WorkerParameters) :
+abstract class FhirSyncWorker(appContext: Context, workerParams: WorkerParameters) :
   CoroutineWorker(appContext, workerParams) {
 
   abstract fun getFhirEngine(): FhirEngine
   abstract fun getDataSource(): DataSource
   abstract fun getSyncData(): ResourceSyncParams
 
   override suspend fun doWork(): Result {
-    // TODO handle retry
     val result = FhirSynchronizer(getFhirEngine(), getDataSource(), getSyncData()).synchronize()
-    if (result is Success) {
-      return Result.success()
+    /**
+     * In case of failure, we can check if its worth retrying and do retry based on
+     * [RetryConfiguration.maxRetries] set by user.
+     */
+    val retries = inputData.getInt(MAX_RETRIES_ALLOWED, 0)
+    return if (result is Success) {
+      Result.success()
+    } else if (retries > runAttemptCount) {
+      Result.retry()
+    } else {
+      Result.failure()
```

Comment:
```suggestion
    return when {
      result is Success -> Result.success()
      retries -> runAttemptCount -> Result.retry()
      else -> Result.failure()
```

would make it easier to read. (feel free to ignore).

Reply:
I'd prefer `when` over `if-else if` since it's easier to read and maintain.

Reply:
Done